Converting schemas from JsonSchema with "oneOf" to Avro/Parquet via Confluent convertors

Hello, 

I am facing an issue with **_JsonSchema schemas with "oneOf" keywords._** 

Following the advices in this[ blog article](https://www.confluent.io/blog/multiple-event-types-in-the-same-kafka-topic/), I am  testing this setup : 
* Sending multiple events types in a single topic, while having a Schema-Registry in place.
* Using json-schema for sending schemas. 
* Using "oneOf" schema as an "umbrella" schema on the topic. The "oneOf" wraps two other schemas. 
* Sink with Kafka-connect, using this format class : "io.confluent.connect.s3.format.parquet.ParquetFormat". 
* (note that the Parquet convertor uses the Avro convertor behind the scenes)

I've started to ingest those messages through a s3-sink in Kafka-Connect, and I eventually want to convert them to Parquet files in s3.  Yet, I have an exception, linked to the **_conversion of the produced schema to Avro Schema._**

It looks like the JsonSchemaConverter outputs schemas with **_fieldnames such as "io.confluent.connect.json.OneOf.field.0",_** which **_clashes with the official avro library_**. So eventually, when such schemas are converted to avro in the AvroData class, it will throw an exception.

=======================

References: 
* the official [avro spec](https://avro.apache.org/docs/current/spec.html#names)
* code line in [JsonSchemaData](https://github.com/confluentinc/schema-registry/blob/master/json-schema-converter/src/main/java/io/confluent/connect/json/JsonSchemaData.java#L932)

https://github.com/confluentinc/schema-registry/blob/b18eedb98546c6c0b385059ba098a602d26ae79f/json-schema-converter/src/main/java/io/confluent/connect/json/JsonSchemaData.java#L932

* function line where it eventually breaks, when we try to create [avro data](https://github.com/confluentinc/schema-registry/blob/b18eedb98546c6c0b385059ba098a602d26ae79f/avro-data/src/main/java/io/confluent/connect/avro/AvroData.java#L1080)

https://github.com/confluentinc/schema-registry/blob/b18eedb98546c6c0b385059ba098a602d26ae79f/avro-data/src/main/java/io/confluent/connect/avro/AvroData.java#L1080

=======================

I've wrapped an isolated small code example that reproduces the issue. 
[QSKafkaConnectSerializationCheck.java.txt](https://github.com/confluentinc/schema-registry/files/6433201/QSKafkaConnectSerializationCheck.java.txt)

I am found this issue while using Confluent 6.1.1 , Avro 1.9.1 ; but it looks like it can apply to latest Confluent and Avro versions too. 

=======================
Stack trace from Kafka-connect logs:
Caused by: org.apache.avro.SchemaParseException: Illegal character in: io.confluent.connect.json.OneOf.field.0
	at org.apache.avro.Schema.validateName(Schema.java:1530)
	at org.apache.avro.Schema.access$400(Schema.java:87)
	at org.apache.avro.Schema$Field.<init>(Schema.java:518)
	at org.apache.avro.Schema$Field.<init>(Schema.java:557)
	at io.confluent.connect.avro.AvroData.addAvroRecordField(AvroData.java:1054)
	at io.confluent.connect.avro.AvroData.fromConnectSchema(AvroData.java:887)
	at io.confluent.connect.avro.AvroData.fromConnectSchema(AvroData.java:710)
	at io.confluent.connect.avro.AvroData.fromConnectSchema(AvroData.java:704)
	at io.confluent.connect.s3.format.parquet.ParquetRecordWriterProvider$1.write(ParquetRecordWriterProvider.java:75)
	at io.confluent.connect.s3.format.KeyValueHeaderRecordWriterProvider$1.write(KeyValueHeaderRecordWriterProvider.java:105)
	at io.confluent.connect.s3.TopicPartitionWriter.writeRecord(TopicPartitionWriter.java:532)
	at io.confluent.connect.s3.TopicPartitionWriter.checkRotationOrAppend(TopicPartitionWriter.java:302)
	at io.confluent.connect.s3.TopicPartitionWriter.executeState(TopicPartitionWriter.java:245)
	at io.confluent.connect.s3.TopicPartitionWriter.write(TopicPartitionWriter.java:196)
	at io.confluent.connect.s3.S3SinkTask.put(S3SinkTask.java:234)
	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:586)

=========

--- a/avro-data/src/main/java/io/confluent/connect/avro/AvroData.java
+++ b/avro-data/src/main/java/io/confluent/connect/avro/AvroData.java
@@ -21,6 +21,9 @@ import com.fasterxml.jackson.databind.node.ArrayNode;
 import com.fasterxml.jackson.databind.node.IntNode;
 import com.fasterxml.jackson.databind.node.JsonNodeFactory;
 import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.util.regex.Pattern;
 import org.apache.avro.JsonProperties;
 import org.apache.avro.generic.GenericData;
 import org.apache.avro.generic.GenericEnumSymbol;
@@ -108,6 +111,9 @@ public class AvroData {
   public static final String AVRO_TYPE_ANYTHING = NAMESPACE + ".Anything";
 
   private static final Map<String, Schema.Type> NON_AVRO_TYPES_BY_TYPE_CODE = new HashMap<>();
+
+  private static Pattern NAME_START_CHAR = Pattern.compile("^[A-Za-z_]");
+  private static Pattern NAME_INVALID_CHARS = Pattern.compile("[^A-Za-z0-9_]");
 
   static {
     NON_AVRO_TYPES_BY_TYPE_CODE.put(CONNECT_TYPE_INT8, Schema.Type.INT8);
@@ -314,6 +320,7 @@ public class AvroData {
   private Cache<AvroSchema, Schema> toConnectSchemaCache;
   private boolean connectMetaData;
   private boolean enhancedSchemaSupport;
+  private boolean scrubInvalidNames;
 
   public AvroData(int cacheSize) {
     this(new AvroDataConfig.Builder()
@@ -330,6 +337,7 @@ public class AvroData {
             avroDataConfig.getSchemasCacheSize()));
     this.connectMetaData = avroDataConfig.isConnectMetaData();
     this.enhancedSchemaSupport = avroDataConfig.isEnhancedAvroSchemaSupport();
+    this.scrubInvalidNames = avroDataConfig.isScrubInvalidNames();
   }
 
   /**
@@ -743,7 +751,7 @@ public class AvroData {
     if (schema.name() != null) {
       String[] split = splitName(schema.name());
       namespace = split[0];
-      name = split[1];
+      name = scrubName(split[1]);
     }
 
     // Extra type annotation information for otherwise lossy conversions
@@ -880,11 +888,12 @@ public class AvroData {
           }
           List<org.apache.avro.Schema.Field> fields = new ArrayList<>();
           for (Field field : schema.fields()) {
+            String fieldName = scrubName(field.name());
             String fieldDoc = schema.parameters() != null
                 ? schema.parameters()
                   .get(AVRO_FIELD_DOC_PREFIX_PROP + field.name())
                 : null;
-            addAvroRecordField(fields, field.name(), field.schema(), fieldDoc, fromConnectContext);
+            addAvroRecordField(fields, fieldName, field.schema(), fieldDoc, fromConnectContext);
           }
           baseSchema.setFields(fields);
         }
@@ -1022,6 +1031,27 @@ public class AvroData {
     }
     fromConnectSchemaCache.put(schema, finalSchema);
     return finalSchema;
+  }
+
+  private String scrubName(String name) {
+    return scrubInvalidNames ? doScrubName(name) : name;
+  }
+
+  // Visible for testing
+  protected static String doScrubName(String name) {
+    try {
+      if (name == null) {
+        return name;
+      }
+      String encoded = URLEncoder.encode(name, "UTF-8");
+      if (!NAME_START_CHAR.matcher(encoded).lookingAt()) {
+        encoded = "x" + encoded;  // use an arbitrary valid prefix
+      }
+      encoded = NAME_INVALID_CHARS.matcher(encoded).replaceAll("_");
+      return encoded;
+    } catch (UnsupportedEncodingException e) {
+      return name;
+    }
   }
 
   public org.apache.avro.Schema fromConnectSchemaWithCycle(

--- a/avro-data/src/main/java/io/confluent/connect/avro/AvroDataConfig.java
+++ b/avro-data/src/main/java/io/confluent/connect/avro/AvroDataConfig.java
@@ -31,6 +31,11 @@ public class AvroDataConfig extends AbstractConfig {
       "Toggle for enabling/disabling enhanced avro schema support: Enum symbol preservation and "
       + "Package Name awareness";
 
+  public static final String SCRUB_INVALID_NAMES_CONFIG = "scrub.invalid.names";
+  public static final boolean SCRUB_INVALID_NAMES_DEFAULT = false;
+  public static final String SCRUB_INVALID_NAMES_DOC =
+      "Whether to scrub invalid names by replacing invalid characters with valid ones";
+
   public static final String CONNECT_META_DATA_CONFIG = "connect.meta.data";
   public static final boolean CONNECT_META_DATA_DEFAULT = true;
   public static final String CONNECT_META_DATA_DOC =
@@ -49,6 +54,8 @@ public class AvroDataConfig extends AbstractConfig {
                 ENHANCED_AVRO_SCHEMA_SUPPORT_DEFAULT,
                 ConfigDef.Importance.MEDIUM,
                 ENHANCED_AVRO_SCHEMA_SUPPORT_DOC)
+        .define(SCRUB_INVALID_NAMES_CONFIG, ConfigDef.Type.BOOLEAN, SCRUB_INVALID_NAMES_DEFAULT,
+                ConfigDef.Importance.MEDIUM, SCRUB_INVALID_NAMES_DOC)
         .define(CONNECT_META_DATA_CONFIG, ConfigDef.Type.BOOLEAN, CONNECT_META_DATA_DEFAULT,
                 ConfigDef.Importance.LOW, CONNECT_META_DATA_DOC)
         .define(SCHEMAS_CACHE_SIZE_CONFIG, ConfigDef.Type.INT, SCHEMAS_CACHE_SIZE_DEFAULT,
@@ -65,6 +72,10 @@ public class AvroDataConfig extends AbstractConfig {
 
   public boolean isConnectMetaData() {
     return this.getBoolean(CONNECT_META_DATA_CONFIG);
+  }
+
+  public boolean isScrubInvalidNames() {
+    return this.getBoolean(SCRUB_INVALID_NAMES_CONFIG);
   }
 
   public int getSchemasCacheSize() {

--- a/avro-data/src/test/java/io/confluent/connect/avro/AvroDataTest.java
+++ b/avro-data/src/test/java/io/confluent/connect/avro/AvroDataTest.java
@@ -264,6 +264,35 @@ public class AvroDataTest {
   }
 
   @Test
+  public void testFromConnectWithInvalidName() {
+    AvroDataConfig avroDataConfig = new AvroDataConfig.Builder()
+        .with(AvroDataConfig.SCRUB_INVALID_NAMES_CONFIG, true)
+        .with(AvroDataConfig.CONNECT_META_DATA_CONFIG, false)
+        .build();
+    AvroData avroData = new AvroData(avroDataConfig);
+    Schema schema = SchemaBuilder.struct()
+        .name("org.acme.invalid record-name")
+        .field("invalid field-name", Schema.STRING_SCHEMA)
+        .build();
+    org.apache.avro.Schema avroSchema = org.apache.avro.SchemaBuilder
+        .record("invalid_record_name").namespace("org.acme") // default values
+        .fields()
+        .requiredString("invalid_field_name")
+        .endRecord();
+    assertThat(avroData.fromConnectSchema(schema), equalTo(avroSchema));
+  }
+
+  @Test
+  public void testNameScrubbing() {
+    assertEquals("abc_2B____", AvroData.doScrubName("abc+-.*_"));
+    assertEquals("abc_def", AvroData.doScrubName("abc-def"));
+    assertEquals("abc_2Bdef", AvroData.doScrubName("abc+def"));
+    assertEquals("abc__def", AvroData.doScrubName("abc  def"));
+    assertEquals("abc_def", AvroData.doScrubName("abc.def"));
+    assertEquals("x0abc_def", AvroData.doScrubName("0abc.def"));
+  }
+
+  @Test
   public void testFromConnectComplex() {
     Schema schema = SchemaBuilder.struct()
         .field("int8", SchemaBuilder.int8().defaultValue((byte) 2).doc("int8 field").build())

--- a/protobuf-converter/src/main/java/io/confluent/connect/protobuf/ProtobufData.java
+++ b/protobuf-converter/src/main/java/io/confluent/connect/protobuf/ProtobufData.java
@@ -26,9 +26,12 @@ import com.google.protobuf.Descriptors.OneofDescriptor;
 import com.google.protobuf.DynamicMessage;
 import com.google.protobuf.Message;
 import com.google.protobuf.util.Timestamps;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
 import java.util.HashSet;
 import java.util.LinkedHashMap;
 import java.util.Set;
+import java.util.regex.Pattern;
 import org.apache.kafka.common.cache.Cache;
 import org.apache.kafka.common.cache.LRUCache;
 import org.apache.kafka.common.cache.SynchronizedCache;
@@ -81,11 +84,15 @@ public class ProtobufData {
       + GOOGLE_PROTOBUF_TIMESTAMP_NAME;
   public static final String GOOGLE_PROTOBUF_TIMESTAMP_LOCATION = "google/protobuf/timestamp.proto";
 
+  private static Pattern NAME_START_CHAR = Pattern.compile("^[A-Za-z]");  // underscore not allowed
+  private static Pattern NAME_INVALID_CHARS = Pattern.compile("[^A-Za-z0-9_]");
+
   private int defaultSchemaNameIndex = 0;
 
   private final Cache<Schema, ProtobufSchema> fromConnectSchemaCache;
   private final Cache<Pair<String, ProtobufSchema>, Schema> toConnectSchemaCache;
   private boolean enhancedSchemaSupport;
+  private boolean scrubInvalidNames;
 
   public ProtobufData() {
     this(new ProtobufDataConfig.Builder().with(
@@ -104,6 +111,7 @@ public class ProtobufData {
     toConnectSchemaCache =
         new SynchronizedCache<>(new LRUCache<>(protobufDataConfig.schemaCacheSize()));
     this.enhancedSchemaSupport = protobufDataConfig.isEnhancedProtobufSchemaSupport();
+    this.scrubInvalidNames = protobufDataConfig.isScrubInvalidNames();
   }
 
   /**
@@ -373,12 +381,14 @@ public class ProtobufData {
     if (cachedSchema != null) {
       return cachedSchema;
     }
-    String name = schema.name();
-    if (name == null) {
-      name = DEFAULT_SCHEMA_NAME + "1";
-    }
+    String fullName = getNameOrDefault(schema.name());
+    String[] split = splitName(fullName);
+    String namespace = split[0];
+    String name = scrubName(split[1]);
+    FromConnectContext ctx = new FromConnectContext();
+    ctx.add(fullName);
     ProtobufSchema resultSchema = new ProtobufSchema(
-        rawSchemaFromConnectSchema(schema).getMessageDescriptor(name)
+        rawSchemaFromConnectSchema(ctx, namespace, name, schema).getMessageDescriptor(name)
     );
     fromConnectSchemaCache.put(schema, resultSchema);
     return resultSchema;
@@ -387,22 +397,17 @@ public class ProtobufData {
   /*
    * DynamicSchema is used as a temporary helper class and should not be exposed in the API.
    */
-  private DynamicSchema rawSchemaFromConnectSchema(Schema rootElem) {
+  private DynamicSchema rawSchemaFromConnectSchema(
+      FromConnectContext ctx, String namespace, String name, Schema rootElem) {
     if (rootElem.type() != Schema.Type.STRUCT) {
       throw new IllegalArgumentException("Unsupported root schema of type " + rootElem.type());
     }
     try {
       DynamicSchema.Builder schema = DynamicSchema.newBuilder();
       schema.setSyntax(ProtobufSchema.PROTO3);
-      String fullName = getNameOrDefault(rootElem.name());
-      String[] split = splitName(fullName);
-      String namespace = split[0];
-      String name = split[1];
       if (namespace != null) {
         schema.setPackage(namespace);
       }
-      FromConnectContext ctx = new FromConnectContext();
-      ctx.add(fullName);
       schema.addMessageDefinition(messageDefinitionFromConnectSchema(ctx, schema, name, rootElem));
       return schema.build();
     } catch (Descriptors.DescriptorValidationException e) {
@@ -425,7 +430,7 @@ public class ProtobufData {
           schema,
           message,
           fieldSchema,
-          field.name(),
+          scrubName(field.name()),
           tag
       );
       if (fieldDef != null) {
@@ -458,7 +463,7 @@ public class ProtobufData {
           schema,
           message,
           field.schema(),
-          field.name(),
+          scrubName(field.name()),
           tag
       );
       if (fieldDef != null) {
@@ -1058,6 +1063,27 @@ public class ProtobufData {
       return fullName.substring(indexLastDot + 1);
     } else {
       return fullName;
+    }
+  }
+
+  private String scrubName(String name) {
+    return scrubInvalidNames ? doScrubName(name) : name;
+  }
+
+  // Visible for testing
+  protected static String doScrubName(String name) {
+    try {
+      if (name == null) {
+        return name;
+      }
+      String encoded = URLEncoder.encode(name, "UTF-8");
+      if (!NAME_START_CHAR.matcher(encoded).lookingAt()) {
+        encoded = "x" + encoded;  // use an arbitrary valid prefix
+      }
+      encoded = NAME_INVALID_CHARS.matcher(encoded).replaceAll("_");
+      return encoded;
+    } catch (UnsupportedEncodingException e) {
+      return name;
     }
   }
 

--- a/protobuf-converter/src/main/java/io/confluent/connect/protobuf/ProtobufDataConfig.java
+++ b/protobuf-converter/src/main/java/io/confluent/connect/protobuf/ProtobufDataConfig.java
@@ -31,6 +31,11 @@ public class ProtobufDataConfig extends AbstractConfig {
       "Toggle for enabling/disabling enhanced protobuf schema support: "
           + "package name preservation";
 
+  public static final String SCRUB_INVALID_NAMES_CONFIG = "scrub.invalid.names";
+  public static final boolean SCRUB_INVALID_NAMES_DEFAULT = false;
+  public static final String SCRUB_INVALID_NAMES_DOC =
+      "Whether to scrub invalid names by replacing invalid characters with valid ones";
+
   public static final String SCHEMAS_CACHE_SIZE_CONFIG = "schemas.cache.config";
   public static final int SCHEMAS_CACHE_SIZE_DEFAULT = 1000;
   public static final String SCHEMAS_CACHE_SIZE_DOC = "Size of the converted schemas cache";
@@ -42,6 +47,8 @@ public class ProtobufDataConfig extends AbstractConfig {
             ENHANCED_PROTOBUF_SCHEMA_SUPPORT_DEFAULT,
             ConfigDef.Importance.MEDIUM,
             ENHANCED_PROTOBUF_SCHEMA_SUPPORT_DOC)
+        .define(SCRUB_INVALID_NAMES_CONFIG, ConfigDef.Type.BOOLEAN, SCRUB_INVALID_NAMES_DEFAULT,
+            ConfigDef.Importance.MEDIUM, SCRUB_INVALID_NAMES_DOC)
         .define(SCHEMAS_CACHE_SIZE_CONFIG,
             ConfigDef.Type.INT,
             SCHEMAS_CACHE_SIZE_DEFAULT,
@@ -56,6 +63,10 @@ public class ProtobufDataConfig extends AbstractConfig {
 
   public boolean isEnhancedProtobufSchemaSupport() {
     return this.getBoolean(ENHANCED_PROTOBUF_SCHEMA_SUPPORT_CONFIG);
+  }
+
+  public boolean isScrubInvalidNames() {
+    return this.getBoolean(SCRUB_INVALID_NAMES_CONFIG);
   }
 
   public int schemaCacheSize() {

--- a/protobuf-converter/src/test/java/io/confluent/connect/protobuf/ProtobufDataTest.java
+++ b/protobuf-converter/src/test/java/io/confluent/connect/protobuf/ProtobufDataTest.java
@@ -1013,6 +1013,33 @@ public class ProtobufDataTest {
   }
 
   @Test
+  public void testFromConnectWithInvalidName() {
+    ProtobufDataConfig protobufDataConfig = new ProtobufDataConfig.Builder()
+        .with(ProtobufDataConfig.SCRUB_INVALID_NAMES_CONFIG, true)
+        .build();
+    ProtobufData protobufData = new ProtobufData(protobufDataConfig);
+    Schema schema = SchemaBuilder.struct()
+        .name("org.acme.invalid record-name")
+        .field("invalid field-name", Schema.STRING_SCHEMA)
+        .build();
+    ProtobufSchema protobufSchema = protobufData.fromConnectSchema(schema);
+    Descriptor descriptor = protobufSchema.toDescriptor();
+    assertEquals("invalid_record_name", descriptor.getName());
+    assertEquals("invalid_field_name", descriptor.getFields().get(0).getName());
+  }
+
+  @Test
+  public void testNameScrubbing() {
+    assertEquals("abc_2B____", ProtobufData.doScrubName("abc+-.*_"));
+    assertEquals("abc_def", ProtobufData.doScrubName("abc-def"));
+    assertEquals("abc_2Bdef", ProtobufData.doScrubName("abc+def"));
+    assertEquals("abc__def", ProtobufData.doScrubName("abc  def"));
+    assertEquals("abc_def", ProtobufData.doScrubName("abc.def"));
+    assertEquals("x0abc_def", ProtobufData.doScrubName("0abc.def"));
+    assertEquals("x_abc_def", ProtobufData.doScrubName("_abc.def"));
+  }
+
+  @Test
   public void testToConnectRecursiveSchema() {
     ProtobufSchema protobufSchema = new ProtobufSchema(
         RecursiveKeyValue.RecursiveKeyValueMessage.getDescriptor());

